Update RuboCop to 1.56.x

archanaserver opened the pull request:
This PR updates the RuboCop version from 1.28.0 to 1.56.0.

#### Requires:

The following PRs were reviewed and adjusted to ensure compatibility with the updated RuboCop/Ruby version:

- [x] https://github.com/theforeman/smart_proxy_omaha/pull/24
  - [x] https://github.com/theforeman/smart_proxy_omaha/pull/25
- [x] https://github.com/theforeman/smart_proxy_dynflow/pull/123
  - [x] https://github.com/theforeman/smart_proxy_dynflow/pull/122
  - [x] https://github.com/theforeman/smart_proxy_dynflow/pull/124
- [x] https://github.com/theforeman/smart_proxy_remote_execution_ssh/pull/101
- [x] https://github.com/theforeman/smart_proxy_salt/pull/83
- [x] https://github.com/theforeman/smart_proxy_monitoring/pull/32
- [x] https://github.com/theforeman/jenkins-jobs/pull/317
   - [x] https://github.com/theforeman/jenkins-jobs/pull/324
     - [x] https://github.com/theforeman/jenkins-jobs/pull/326

For the rest of the smart-proxy plugins (`ansible`, `openscap`, `dhcp_infoblox`, `dhcp_remote_isc`, `discovery`, `dns_infoblox`, `pulp`), no changes were required to support Ruby v3.

archanaserver committed:
Add additional cops to rubocop-todo.yaml

Added additional RuboCop cops and updated .rubocop_todo.yml
to address various code quality issues in the smart-proxy project.
The changes help improve code consistency.

ekohl commented:
FYI: there is https://projects.theforeman.org/issues/35374 open as an issue. There I chose version 1.35.x. 1.28.x is the latest version to support Ruby 2.5, but if we move to 2.7+ I think it's safe to go with a newer version. At the time that was 1.35.x, but it's probably even safe to move to 1.48.x which is the current latest.

ehelms commented:
@archanaserver When you get time, can you come back around to this and do the update to 1.35 (or if there is a newer latest) ?

ekohl commented on bundler.d/test.rb:
```diff
@@ -4,7 +4,7 @@
   gem 'benchmark-ips'
   gem 'ruby-prof', '< 1.4'
-  gem 'rubocop', '~> 0.89.0'
+  gem 'rubocop', '~> 1.28.0'
```
We can update to the latest version:
```suggestion
  gem 'rubocop', '~> 1.55.0'
```

archanaserver committed:
Update RuboCop version to 1.56.0

Bumped the RuboCop version from 1.28.0 to 1.56.0 to ensure the
project benefits from the latest improvements, bug fixes, and
code analysis features provided by the updated RuboCop release.

archanaserver committed:
Fix Rubocop Lint/RedundantStringCoercion

archanaserver committed:
Fix Rubocop Style/SafeNavigation

archanaserver committed:
Disable Style/RedundantCondition

archanaserver commented:
@ekohl, could you please review this?

ekohl commented on .rubocop.yml:
```diff
@@ -137,3 +137,6 @@ Style/TrailingCommaInHashLiteral:
 Style/WordArray:
   Enabled: false
+
+Style/RedundantCondition:
+  Enabled: false
```
I would not disable this cop just for one thing that is hard to rewrite. Instead I'd fix the particular case. You can write:
```ruby
type = params.fetch(:type) { name.end_with?('.in-addr.arpa', '.ip6.arpa') ? 'PTR' : 'A' }
```

ekohl reviewed (changes requested):
I'd also enable `rubocop-rake` as a plugin.

archanaserver committed:
Fix Rubocop Style/RedundantCondition

archanaserver committed:
Enable rubocop-rake plugin

archanaserver committed:
Fix Rubocop Rake/Desc

ekohl commented on Gemfile:
```diff
@@ -3,6 +3,7 @@ source 'https://rubygems.org'
 gemspec
 
 gem 'concurrent-ruby', '~> 1.0', require: 'concurrent'
+gem 'rubocop-rake'
```
Please keep it in the right place: If you place it here then it will also be required in production which isn't true for `bundler.d/test.rb`.

ekohl commented on tasks/jenkins.rake:
```diff
@@ -3,17 +3,24 @@ require 'English'
       task :test_unit => [:pre_ci, "ci:setup:testunit"]
     end
 
+    desc 'Run RuboCop analysis'
```
We no longer use this task so it's better to drop it. This is how we really invoke it now:

https://github.com/theforeman/jenkins-jobs/blob/a44d192c1549be46b3f0d85c254f4247c4ec9331/theforeman.org/pipelines/test/smart-proxy.groovy#L47

Note that if you drop this task, you no longer need the `require 'English'` at the top.

ekohl commented on tasks/jenkins.rake:
```diff
@@ -3,17 +3,24 @@ require 'English'
-      task :minitest  => [:pre_ci, "ci:setup:minitest"]
+
+      desc 'Sets up pre-CI environment for MiniTest'
+      task :minitest => [:pre_ci, "ci:setup:minitest"]
```
This repository doesn't use minitest so we can drop this. It's a left over from when we copied it over to multiple repositories.

ekohl commented on tasks/jenkins.rake:
```diff
@@ -3,17 +3,24 @@ require 'English'
     namespace :setup do
+      desc 'Sets up pre-CI environment for Minitest'
       task :pre_ci do
         ENV["CI_REPORTS"] = 'jenkins/reports/unit/'
```
While you're at it, I think you can actually drop this. We now set it in CI: https://github.com/theforeman/jenkins-jobs/blob/a44d192c1549be46b3f0d85c254f4247c4ec9331/theforeman.org/pipelines/test/smart-proxy.groovy#L97

ekohl commented on tasks/jenkins.rake:
```diff
@@ -3,17 +3,24 @@ require 'English'
+      desc 'Sets up pre-CI environment for Minitest'
       task :pre_ci do
         ENV["CI_REPORTS"] = 'jenkins/reports/unit/'
         gem 'ci_reporter'
```
I wonder if this is still needed. We have an explicit require on line 4 already. I think the whole `pre_ci` task can be dropped.

ekohl commented on tasks/jenkins.rake:
```diff
@@ -3,17 +3,24 @@ require 'English'
   require "ci/reporter/rake/test_unit"
   namespace :jenkins do
+    desc 'Sets up pre-CI environment for testing'
     task :unit => ["jenkins:setup:test_unit", 'rake:test']
```
You can probably drop most tasks and reduce this to
```suggestion
    task :unit => ['ci:setup:testunit', 'rake:test']
```

ekohl commented on tasks/jenkins.rake:
```diff
@@ -3,17 +3,24 @@ require 'English'
+      task :minitest => [:pre_ci, "ci:setup:minitest"]
+
+      desc 'Sets up pre-CI environment for Test::Unit'
       task :test_unit => [:pre_ci, "ci:setup:testunit"]
```
If you're dropping `pre_ci`, this also becomes redundant.

ekohl commented on tasks/jenkins.rake:
```diff
@@ -3,17 +3,24 @@ require 'English'
 begin
   require "ci/reporter/rake/test_unit"
   namespace :jenkins do
+    desc 'Sets up pre-CI environment for testing'
```
This doesn't just set it up: it requires both the setup and then runs the tests.

ekohl commented on .rubocop_todo.yml:
```diff
@@ -1,280 +1,551 @@
-# `rubocop --auto-gen-config --exclude-limit 0`
-# on 2020-02-26 21:20:38 +0100 using RuboCop version 0.80.0.
+# `rubocop --auto-gen-config --no-offense-counts --no-auto-gen-timestamp`
+# using RuboCop version 1.28.2.
```
Looks like you didn't update this with 1.55.x.

ekohl commented on bundler.d/test.rb:
```diff
@@ -4,7 +4,7 @@
   gem 'benchmark-ips'
   gem 'ruby-prof', '< 1.4'
-  gem 'rubocop', '~> 0.89.0'
+  gem 'rubocop', '~> 1.55.0'
```
By now 1.56 is also out.
```suggestion
  gem 'rubocop', '~> 1.56.0'
```

archanaserver commented:
@ekohl the changes is updated now, as per the comments. Thanks for the review!

ekohl commented on bundler.d/test.rb:
```diff
@@ -4,11 +4,12 @@
   gem 'rubocop-performance', '~> 1.5.2'
   gem 'rack-test'
   gem 'rake'
   gem 'rubocop-checkstyle_formatter', '~> 0.2'
```
We actually no longer use this so it can be dropped
```suggestion
```

ekohl commented on tasks/jenkins.rake:
```diff
@@ -1,28 +1,7 @@
-require 'English'
-
 begin
```
If you don't catch the `LoadError` you don't need this `begin` and `end`.

archanaserver commented:
@ekohl I looked into the failure but can't find the issue, can you help what is the issue here? https://ci.theforeman.org/blue/organizations/jenkins/smart-proxy-pr-test/detail/smart-proxy-pr-test/275/pipeline

ekohl commented on modules/dns/dns_api.rb:
```diff
@@ -47,11 +47,7 @@ class Api < ::Sinatra::Base
-      else
-        type = (name =~ /\.(in-addr|ip6)\.arpa$/) ? "PTR" : "A"
-      end
+      type = params.fetch(:type) { name.end_with?('.in-addr.arpa', '.ip6.arpa') ? 'PTR' : 'A' }
```
This fails because `params` does contain the key `type` but has the value `nil`. `fetch` only works if the key isn't present. So this isn't going to work.

A simpler way may be
```suggestion
      type = params[:type]
      type ||= name.end_with?('.in-addr.arpa', '.ip6.arpa') ? 'PTR' : 'A'
```

ekohl commented on bundler.d/test.rb:
```diff
@@ -4,11 +4,11 @@
   gem 'rack-test'
   gem 'rake'
-  gem 'rubocop-checkstyle_formatter', '~> 0.2'
+  gem 'rubocop-rake'
```
I know it isn't completely sorted, but I'd at least make sure the rubocop dependencies are sorted.
